cpp/src/io/parquet/parquet.hpp

```diff
@@ -362,17 +362,111 @@ class CompactProtocolWriter {
   }
 
  public:
-#define DECL_CPW_STRUCT(st) size_t write(const st *)
-  DECL_CPW_STRUCT(FileMetaData);
-  DECL_CPW_STRUCT(SchemaElement);
-  DECL_CPW_STRUCT(RowGroup);
-  DECL_CPW_STRUCT(KeyValue);
-  DECL_CPW_STRUCT(ColumnChunk);
-  DECL_CPW_STRUCT(ColumnChunkMetaData);
-#undef DECL_CPW_STRUCT
+  size_t write(const FileMetaData *);
+  size_t write(const SchemaElement *);
+  size_t write(const RowGroup *);
+  size_t write(const KeyValue *);
+  size_t write(const ColumnChunk *);
+  size_t write(const ColumnChunkMetaData *);
 
  protected:
   std::vector<uint8_t> *m_buf;
+
+  friend class CompactProtocolWriterBuilder;
+};
+
+class CompactProtocolWriterBuilder {
+  CompactProtocolWriter *ptr;
+  size_t struct_start_pos;
+  int current_field;
+
+ public:
+  CompactProtocolWriterBuilder(CompactProtocolWriter *cpw_ptr)
```

Comment:
The name's not quite right. It suggests that the class is building `CompactProtocolWriter` objects. Maybe something like `compact_protocol_field_writer`?